src/training/finetune_glue.py

```diff
@@ -0,0 +1,284 @@
+        except ImportError:
+            raise ImportError("Please install HF datasets: pip install datasets")
+
+        self.dataset = load_dataset("glue", task, split=split)
+        self.label_key = label_key
+        self.text_key = text_key
+        self.length = len(self.dataset)
+        self.tokenize = tokenizer
+
+    def __len__(self):
+        return self.length
+
+    def __getitem__(self, idx):
+        item = self.dataset[idx]
+        text = item[self.text_key]
+        label = item[self.label_key]
+        return {
+            'text': self.tokenize([text])[0],
+            'label': label
+        }
+
+
+class TextClassifier(nn.Module):
+    def __init__(self, encoder, embed_dim, num_labels):
+        super().__init__()
+
+        self.encoder = encoder
+        self.logits_proj = nn.Linear(embed_dim, num_labels)
+
+    def forward(self, text):
+        text_features = self.encoder.encode_text(text)
+        logits = self.logits_proj(text_features)
+        return logits
+
+
+def get_task_metric(task_name):
+    if task_name == "cola":
+        metric = evaluate.load("glue", "stsb")
+    else:
+        metric = evaluate.load("glue", task_name)
+    return metric
+
+
+def get_task_dataloaders(args):
+    tokenizer = get_tokenizer(args.model)
+    task_name = args.task_name
+
+    dataloaders = {}
+    for split_name in ["train", "validation", "test"]:
+        dataset = GLUEDataset(
+            task_name,
+            split_name,
+            text_key="sentence",
+            label_key="label",
+            tokenizer=tokenizer,
+        )
+        dataloader = DataLoader(
+            dataset,
+            batch_size=args.batch_size,
+            shuffle=True,
+            num_workers=args.workers,
+            pin_memory=True,
+            drop_last=True,
+        )
+        dataloaders[split_name] = dataloader
+
+    return dataloaders
+
+
+def compute_metrics(model, dataloader, device, args):
+    model.eval()
+    metric = get_task_metric(args.task_name)
+    val_loss = 0
+    samples_seen = 0
+    for batch in dataloader:
+        with torch.no_grad():
+            text = batch["text"].to(device)
+            label = batch["label"].to(device)
+            samples_seen += text.shape[0]
+            logits = model(text)
+            logits = logits.view(-1)
+            label = label.view(-1).float()
+            predictions = torch.sigmoid(logits) > 0.5
+            batch_val_loss = nn.functional.binary_cross_entropy_with_logits(logits, label, reduction='sum')
+        val_loss += batch_val_loss.item()
+        metric.add_batch(
+            predictions=predictions.cpu().numpy(),
+            references=label.cpu().numpy(),
+        )
+    model.train()
+    metrics = metric.compute()
+    metrics["loss"] = val_loss / samples_seen
+    return metrics
+
+
+def train_one_epoch(model, data, epoch, optimizer, scheduler, early_stop, device, args):
+    model.train()
+    progress_bar = tqdm(total=len(data["train"]))
```

Comment:
When you are using distributed data sampler. You also need to call `set_epoch` on the dataloader so as to make sure each epoch is loading data in different order.